### ISSUE TITLE
Engine linking and `libcrypto` interface policy

### ISSUE THREAD

**Issue by yanovich** [openssl/openssl#17508]
If any executable loads an engine, the executable already either has `libcrypto.so` loaded or is statically linked against `libcrypto.a`. Anyway it already has all `libcrypto` symbols present.

All in-tree engines are linked against `libcrypto.so` now. As a result, a diamond dependency is introduced. When such an engine is then loaded by an executable which is statically linked against `libcrypto.a`, `ld` will insist on loading `libcripto.so`, despite the executable already having all necessary symbols. When the executable is statically linked, shared object for `libcrypto.so` is usually not built, and `ld` won't find it.

Diamond dependency also allows `libcrypto.so.3` to load engines built against `libcrypto.so.1.1`, and vice versa. When the engine is used the executable segfaults. I've faced this issue when experimenting with `node.js-v17` which uses `openssl-3`, and `gost-engine`. Removing link-time dependency on `libcrypto.so` [fixes](https://github.com/gost-engine/engine/pull/384) diamond dependency problem for `gost-engine`. I will create a pull request with the proposed fix for in-tree engines.

The discussion with `gost-engine` maintainers showed that the problem has a second aspect. If `libcrypto` modifies any of its exported interfaces (structures and function signatures), the engine will still load, but may encounter hard-to-debug issues. So the second question is whether a policy on `libcrypto` exports can be enforced, so that changed to their signatures are never changed, only added or removed. That is if `a(b, c)` need to become `a(b, c, d)`, then instead `a(b, c)` is removed and `a2(b, c, d)` is added.

**Comment by yanovich** [comment]
@beldmit, @vt-alt

**Comment by levitte** [comment]
Engines *must* be linked with the same libcrypto as the calling application, for exactly the reasons you mention. Anything else is an undefined scenario.

**Comment by levitte** [comment]
Or yeah, not linking it at all with libcrypto and relying om the symbols to already be present works too... on platforms that support this, which not all do.

**Comment by yanovich** [comment]
> Engines _must_ be linked with the same libcrypto as the calling application, for exactly the reasons you mention. Anything else is an undefined scenario.

Yep. And I propose a way to ensure this exact behavior on systems that have `ld` linker.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of the 3.4 dev cycle, barring further input
